### PR TITLE
chore: Fix bugs in the index-builders

### PR DESCRIPTION
**Pull request by benclive**
**What this PR does / why we need it**:
Fixes a few bugs with the index-builder:
* Concurrent map access inside the builder after introducing parallelism in reading the stream sections
* Correctly reading tenant info from `section.Tenant` instead of `logsSection.Tenant()`, as the latter is empty (I will remove that method in another PR)
* Removing functionality to flush Pointers & Stream sections on hitting the max section size. Stream sections in log objects already do this, and Pointers sections need to maintain state about which sections have been seen to avoid duplicating data. This was never a problem for real data as they didn't get near the section size, but I picked it up as an edge case in tests.
* Fixes an edge case where a builder would fail to flush if a tenant's builder was empty.

**Which issue(s) this PR fixes**:
Part of https://github.com/grafana/loki-private/issues/1769

### CODE REVIEW

File: pkg/dataobj/index/calculate.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ func (c *Calculator) Flush() (*dataobj.Object, io.Closer, error) {
 // Calculate reads the log data from the input logs object and appends the resulting indexes to calculator's builder.
 // Calculate is not thread-safe.
 func (c *Calculator) Calculate(ctx context.Context, logger log.Logger, reader *dataobj.Object, objectPath string) error {
+	fmt.Println("Calculate", objectPath)
```
Comment (chaudum) [inline pkg/dataobj/index/calculate.go]: Please remove or convert to log.
